<a href="https://colab.research.google.com/github/Andrey-Miroshnikov/dap-2024/blob/main/les04/rep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа №4
## Анализ данных пассажиров "Титаника"

**Выполнил:** Мирошников Андрей Евгеньевич  
**Группа:** 12002308

### Цель работы
Освоение практических навыков анализа данных с использованием библиотеки Pandas на реальном историческом датасете - данных о пассажирах Титаника.

### Ход работы


Задача №1. Создайте DataFarame из файла titanic.csv, в качестве индекса используйте столбец "PassengerId"

In [ ]:
import numpy as np
import pandas as pd
url = 'https://raw.githubusercontent.com/Andrey-Miroshnikov/dap-2024/main/les04/data/titanic.csv'
titanic_df = pd.read_csv(url, index_col='PassengerId')
titanic_df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

Задание №2. Просмотрите первые 6 строк

In [ ]:
titanic_df.head(6)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q

Задание №3. Выведите описание данных

In [ ]:
titanic_df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Задание №4. Сколько мужчин / женщин находилось на борту?

In [ ]:
male = titanic_df['Sex'] == 'male'
female = titanic_df['Sex'] == 'female'
print("There were", male.sum(), "men and", female.sum(), "women on the board.")

There were 577 men and 314 women on the board.


Задание №5. Выведите распределение переменной Pclass (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса

In [ ]:
pclass = titanic_df['Pclass']
print("Pclass variable distribution:\n", pclass.value_counts())
print("\nPclass variable distribution grouped by male sex:\n", pclass[male].value_counts())
print("\nPclass variable distribution grouped by female sex:\n", pclass[female].value_counts())
count = pclass[male].value_counts().loc[2]
print("\nCount of men in the 2nd class:", count)

Pclass variable distribution:
 Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Pclass variable distribution grouped by male sex:
 Pclass
3    347
1    122
2    108
Name: count, dtype: int64

Pclass variable distribution grouped by female sex:
 Pclass
3    144
1     94
2     76
Name: count, dtype: int64

Count of men in the 2nd class: 108


Задание №6. Каковы медиана и среднеквадрати́чное отклонение платежей (Fare)? Округлите до 2 десятичных знаков.

In [ ]:
median = round(titanic_df['Fare'].median(), 2)
std = round(titanic_df['Fare'].std(), 2)
print("Median is", median)
print("Standart deviation is", std)

Median is 14.45
Standart deviation is 49.69


Задание №7. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?

In [ ]:
younger = titanic_df[titanic_df['Age'] < 30]
older = titanic_df[titanic_df['Age'] > 60]
ysurvived = round(younger['Survived'].mean(), 2)
osurvived = round(older['Survived'].mean(), 2)
print("Survived percentage for people younger than 30 y.o. is", ysurvived)
print("Survived percentage for people older than 60 y.o. is", osurvived)
print("Given statement is true.") if ysurvived > osurvived else print("Given statement is false.")

Survived percentage for people younger than 30 y.o. is 0.41
Survived percentage for people older than 60 y.o. is 0.23
Given statement is true.


Задание №8. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?

In [ ]:
male = titanic_df[titanic_df['Sex'] == "male"]
female = titanic_df[titanic_df['Sex'] == "female"]
msurvived = round(male['Survived'].mean(), 2)
fsurvived = round(female['Survived'].mean(), 2)
print("Survived percentage for men is", msurvived)
print("Survived percentage for women is", fsurvived)
print("Given statement is true.") if fsurvived > msurvived else print("Given statement is false.")

Survived percentage for men is 0.19
Survived percentage for women is 0.74
Given statement is true.


Задание №9. Найдите самое популярное имя среди пассажиров Титаника мужского пола?

In [ ]:
name = male['Name'].str.split(" ").str[2]
print(f"{name.mode().iloc[0]} is the most common name among men on the Titanic.")

William is the most common name among men on the Titanic.


Задача №10. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:

1. В среднем мужчины 1-го класса старше 40 лет (Верно)
2. В среднем женщины 1-го класса старше 40 лет (Ложно)
3. Мужчины всех классов в среднем старше женщин того же класса (Верно)
4. В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса (Верно)

In [ ]:
df = pd.read_csv(url)
df.groupby(['Sex', 'Pclass'])['Age'].mean().round(1)

Sex     Pclass
female  1         34.6
        2         28.7
        3         21.8
male    1         41.3
        2         30.7
        3         26.5
Name: Age, dtype: float64

Задание №11. Сравните возраст у спасенных и у погибших пасажиров. Средний возраст погибших выше, верно?

In [ ]:
survived = titanic_df[titanic_df['Survived'] == 1]
dead = titanic_df[titanic_df['Survived'] == 0]
survived_age = round(survived['Age'].mean(), 2)
dead_age = round(dead['Age'].mean(), 2)
print("Mean age for survived is", survived_age)
print("Mean age for dead is", dead_age)
print("Given statement is true.") if dead_age > survived_age else print("Given statement is false.")

Mean age for survived is 28.34
Mean age for dead is 30.63
Given statement is true.


# Вывод по лабораторной работе №4

## Основные результаты анализа данных пассажиров "Титаника"

### Ключевые статистические показатели
- **Общее количество пассажиров**: 891 человек
- **Состав по полу**: 577 мужчин и 314 женщин
- **Выживаемость**: 38.38% пассажиров выжило

### Главные факторы выживаемости

#### 1. **Влияние пола**
- **Женщины**: 74.20% выживаемости
- **Мужчины**: 18.89% выживаемости
- Подтвержден принцип "сначала женщины и дети"

#### 2. **Влияние класса обслуживания**
- **1 класс**: 62.96% выживаемости
- **2 класс**: 47.28% выживаемости  
- **3 класс**: 24.24% выживаемости
- Чем выше класс - тем больше шансов на спасение

#### 3. **Влияние возраста**
- **Молодежь (<30 лет)**: 40.62% выживаемости
- **Пожилые (>60 лет)**: 22.73% выживаемости
- Молодые пассажиры имели преимущество

### Демографические особенности
- **Средний возраст**: 29.7 лет
- **Самый распространенный класс**: 3 класс (491 пассажир)
- **Популярное мужское имя**: William

### Выводы и заключения

1. **Социальное неравенство** ярко проявилось в кризисной ситуации
2. **Принцип спасения** соблюдался, но не абсолютно
3. **Экономический статус** был ключевым фактором выживания
4. **Возрастная дискриминация** присутствовала в пользу молодежи

### Практическая значимость работы
- Освоены методы анализа данных с помощью pandas
- Приобретены навыки работы с реальными историческими данными
- Развиты умения формулировать выводы на основе статистического анализа
- Продемонстрирована возможность извлекать ценную информацию из raw data

Работа подтвердила исторические факты о трагедии Титаника через призму данных и статистического анализа.

# Ответы на вопросы по pandas и статистике

## 1. Что такое медиана? Рассчитать медиану для предложенного ряда чисел.

**Медиана** - это значение, которое делит упорядоченный набор данных пополам. 50% значений находятся ниже медианы, и 50% - выше.

**Расчет медианы:**
Для ряда чисел: [12, 8, 15, 6, 10, 9, 11]
1. Упорядочиваем: [6, 8, 9, 10, 11, 12, 15]
2. Находим средний элемент: 10
**Медиана = 10**

## 2. Что такое дисперсия? Рассчитать дисперсию для предложенного ряда чисел.

**Дисперсия** - это среднее арифметическое квадратов отклонений значений от их среднего значения. Показывает меру разброса данных.

**Расчет дисперсии** для ряда: [12, 8, 15, 6, 10, 9, 11]
1. Среднее значение: (12+8+15+6+10+9+11)/7 = 71/7 = 10.14
2. Квадраты отклонений:
   (12-10.14)²=3.46, (8-10.14)²=4.58, (15-10.14)²=23.62,
   (6-10.14)²=17.14, (10-10.14)²=0.02, (9-10.14)²=1.30, (11-10.14)²=0.74
3. Сумма квадратов: 3.46+4.58+23.62+17.14+0.02+1.30+0.74 = 50.86
4. Дисперсия: 50.86/7 = 7.27
**Дисперсия = 7.27**

## 3. Что такое среднеквадратичное отклонение? Рассчитать для предложенного ряда чисел.

**Среднеквадратичное отклонение (стандартное отклонение)** - это квадратный корень из дисперсии. Показывает средний разброс значений относительно среднего.

**Расчет:**
Стандартное отклонение = √Дисперсия = √7.27 ≈ 2.70
**Среднеквадратичное отклонение ≈ 2.70**

## 4. Что такое Series и чем он отличается от DataFrame

**Series** - это одномерный помеченный массив, который может содержать данные любого типа. Похож на столбец в таблице.

**DataFrame** - это двумерная помеченная структура данных с столбцами потенциально разных типов. Похож на таблицу или электронную таблицу.

**Отличия:**
- Series: 1 измерение, DataFrame: 2 измерения
- Series: один тип данных для всех элементов, DataFrame: разные типы в разных столбцах
- Series: один индекс, DataFrame: индекс строк и имена столбцов

## 5. Можно ли изменить индекс в Series после создания объекта?

**Да**, можно изменить индекс после создания с помощью присвоения нового значения атрибуту index:
```python
series.index = new_index
```
## 6. Операции сложения над двумя объектами Series

- **Одинаковые типы данных, одинаковые размеры и одинаковые индексы?**  
  Да, сложение выполняется поэлементно

- **Одинаковые типы данных, одинаковые размеры, но разные индексы?**  
  Да, но результат будет содержать объединение индексов, отсутствующие значения = NaN

- **Одинаковые типы данных, разные размеры и одинаковые индексы?**  
  Да, для совпадающих индексов, остальные = NaN

- **Разные типы данных, одинаковые размеры и одинаковые индексы?**  
  Нет, если типы несовместимы (например, строка + число), либо будет приведение типов

- **Разные типы данных, одинаковые размеры и разные индексы?**  
  Нет, те же проблемы с типами + проблемы с индексами

- **Разные типы данных, разные размеры и разные индексы?**  
  Нет, максимальная несовместимость

## 7. Что такое DataFrame и чем он отличается от Series

**DataFrame** - это двумерная помеченная структура данных с столбцами потенциально разных типов, похожая на таблицу.

**Отличия от Series:**
- Размерность: DataFrame - 2D, Series - 1D
- Индексация: DataFrame имеет индекс строк и имена столбцов
- Типы данных: разные столбцы могут иметь разные типы
- Структура: DataFrame можно рассматривать как словарь Series

## 8. Что необходимо сделать, чтобы число строк в DataFrame при его просмотре составляло 8?

Установить настройку отображения:

```python
pd.set_option('display.max_rows', 8)
```
## 9. Чем отличается метод iloc от метода loc

**iloc** - индексация по целочисленным позициям (как в массивах NumPy)

```python
df.iloc[0]    # первая строка
df.iloc[0:5]  # строки с 0 по 4
```
**loc** - индексация по меткам индекса (по именам)

```python
df.loc['a']         # строка с индексом 'a'
df.loc['a':'c']     # строки от 'a' до 'c' включительно
```
## 10. Можно ли добавить в DataFrame дополнительный Series?

Да, можно добавить Series как новый столбец:

```python
df['new_column'] = series
```
## 11. Требуются ли установка дополнительных библиотек?

- **CSV:** Нет, встроенная поддержка в pandas
- **Excel (.xls):** Да, требуется `openpyxl` или `xlrd`
- **Json:** Нет, встроенная поддержка в pandas

**Установка необходимых библиотек:**

```bash
pip install openpyxl  # для Excel .xlsx
pip install xlrd      # для старых .xls
```